internal/ingress/controller/handlers/pod.go

```diff
@@ -0,0 +1,118 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ handler.EventHandler = (*EnqueueRequestsForPodsEvent)(nil)
+
+type EnqueueRequestsForPodsEvent struct {
+	IngressClass string
+	Cache        cache.Cache
+}
+
+// Create is called in response to an create event - e.g. Pod Creation.
+func (h *EnqueueRequestsForPodsEvent) Create(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
+	h.enqueueImpactedIngresses(e.Object.(*corev1.Pod), queue)
+}
+
+// Update is called in response to an update event -  e.g. Pod Updated.
+func (h *EnqueueRequestsForPodsEvent) Update(e event.UpdateEvent, queue workqueue.RateLimitingInterface) {
+	podOld := e.ObjectOld.(*corev1.Pod)
+	podNew := e.ObjectNew.(*corev1.Pod)
+	if !reflect.DeepEqual(podOld, podNew) {
+		h.enqueueImpactedIngresses(podNew, queue)
+	}
+}
+
+// Delete is called in response to a delete event - e.g. Pod Deleted.
+func (h *EnqueueRequestsForPodsEvent) Delete(e event.DeleteEvent, queue workqueue.RateLimitingInterface) {
+}
+
+// Generic is called in response to an event of an unknown type or a synthetic event triggered as a cron or
+// external trigger request - e.g. reconcile Autoscaling, or a Webhook.
+func (h *EnqueueRequestsForPodsEvent) Generic(event.GenericEvent, workqueue.RateLimitingInterface) {
+}
+
+func (h *EnqueueRequestsForPodsEvent) enqueueImpactedIngresses(pod *corev1.Pod, queue workqueue.RateLimitingInterface) {
+	ingressList := &extensions.IngressList{}
+	if err := h.Cache.List(context.Background(), client.InNamespace(pod.Namespace), ingressList); err != nil {
+		glog.Errorf("failed to fetch ingresses impacted by pod %s due to %v", pod.GetName(), err)
+		return
+	}
+
+	if pod.Status.PodIP == "" {
+		return
+	}
+
+	for _, ingress := range ingressList.Items {
+		if !class.IsValidIngress(h.IngressClass, &ingress) {
+			continue
+		}
+
+		backends, _, err := tg.ExtractTargetGroupBackends(&ingress)
+		if err != nil {
+			glog.Errorf("failed to extract backend services from ingress %s/%s, reconciling the ingress. Error: %e",
+				ingress.Namespace, ingress.Name, err)
+			queue.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: ingress.Namespace,
+					Name:      ingress.Name,
+				},
+			})
+			break
+		}
+
+		for _, backend := range backends {
+			endpoint := &corev1.Endpoints{}
+			nspname := types.NamespacedName{
+				Namespace: ingress.Namespace,
+				Name:      backend.ServiceName,
+			}
+			if err = h.Cache.Get(context.Background(), nspname, endpoint); err != nil {
+				glog.Errorf("failed to fetch enpoint %s backing ingress %s/%s, ignoring",
+					backend.ServiceName, ingress.Namespace, ingress.Name)
+				continue
+			}
+
+			if h.isPodInEndpoint(pod, endpoint) {
```

Comment:
Would it make sense to only reconcile the ingress if the `ContainersReady` condition is true to avoid unnecessary reconciliation iterations (and thus many AWS API calls) on frequent pod status updates during pod startup?

Reply:
@devkid the point of reconciliations IMO is that you don't have to rely on state comparison to make a decision. 

In our test we didn't had any issues yet with AWS API limits.

Also we shouldn't rely on the fact that right now only containers ready is the important and driving factor for registering pods in the endpoints table. Upstream k8s could change or extend that logic any time and it would break the implementation.

Reply:
I agree with @nirnanaaa, that would specialise the watch by embedding expectations on implementations details of an unrelated part in the code base (which might change); eg. setting a trap for futur maintainers.

Reply:
@nirnanaaa 
I think we need to have the ContainerReady check. we have the check to decide whether to register the Pod to ALB anyway. Under current architecture, the reconcile will call AWS APIs to check every ALB settings before reconcile targetGroups, which is inefficient and cause troubles if there are many ingresses.

I think it's better to have below changes:
1. have a common function defined in backends package like "IsPodSuitableAsIPTarget", which checks containersReady
2. change https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/internal/ingress/backend/endpoint.go#L186 to call the function above for decide whether pod in NotReadyAddresses should be registered.
3. only enqueue Ingress for pod Update event, when PodInEndpoint(podNew)  && IsPodSuitableAsIPTarget(podNew) != IsPodSuitableAsIPTarget(podOld)

Reply:
@M00nF1sh does that really justify adding all the complex logic to have oldRef stored, the handling if no old ref was found in store etc ? I don't think the comparison would be a huge deal other than duplicating logic.

Reply:
Usually I've seen that not that many pod updates are triggered other than when the containers are ready. We're currently running this patch in prod.

Reply:
@nirnanaaa i think it helps to reduce reconcile calls under current architecture. and podOld and podNew are guarantee to exists(by informer). It's a optimization other than requirement.
In the future, we can let user control the "IsPodSuitableAsIPTarget" logic, e.g. check for additional readinessGates before register pod to ALB.

Reply:
Understood, thanks for clarifying. Would it might make sense to first check if that's an actual problem ? Theoretically there aren't that many pod events happening and we would save time by not overengineering a non existing problem.

Reply:
Understood, thanks for clarifying. Would it might make sense to first check if that's an actual problem ? Theoretically there aren't that many pod events happening and we would save time by not overengineering a non existing problem.

Reply:
@nirnanaaa I already got a lot issues caused by the controller generating too many reconcile s and cause AWS api throttling :D. (and that's my motivation to separate targetReconcile from ALB reconcile in v2).

Reply:
Fully understood. Sorry if I sound rough -  I just want to find the best possible solution. The ingress controller codebase is already full of checks left and right. So keeping a clean base is also something important ;)

Reply:
@nirnanaaa :D sure, i think it's pretty clean too, and makes the precondition of whether to add pod into ALB explicit. :D

Reply:
@M00nF1sh pushed a commit doing exactly that

Reply:
thanks, ready to go :D, just  one nit, we need a "break" after queue.Add :D

Reply:
Thanks, added the `break` (optimize the loop but precludes pods being part of several readiness gate backed ingresses - which I guess was a shady corner case wrt readiness gates semantics anyway?).

Reply:
@bpineau The break would only quit the loop that iterates over the backends, i.e. all other ingresses would still be processed. Enqueuing the ingress for reconciliation will reconcile all target groups anyhow.

Reply:
@devkid you're right